### PR TITLE
DG-1103

### PR DESCRIPTION
**Pull request by restevens402**
So in the code, we set the amount deducted for the rate limit in the cache.  Problem is, that's a single key value for the Tx.hash.  Because we are also deducting from the receiver, we were overwriting the "amount" for that transaction based on the minimum of the receiver balance vs the hertz used ( so zero balance for a contract=zero deduction!!!! ).   So the tx.hash rate limit "amount" was being set back to zero.  So when we do the deduction for each tx.hash that is still under the TTL in the cache, the deduction for that tx.hash was zero.

### CODE REVIEW

File: commons/types/rate_limit.go [hunk 1]
```diff
@@ -62,6 +62,13 @@ func (this *RateLimit) Set(window Window, txn *badger.Txn, cache *cache.Cache) e
 
 func (this *RateLimit) cache(window Window, cache *cache.Cache) {
 	cache.Set(getAccountRateLimitKey(this.Address), this.Existing, TransactionCacheTTL)
+	utils.Debug("Current Window TTL = ", window.TTL)
```
Comment (brianosaurus) [inline commons/types/rate_limit.go]: @restevens402 Do we have anything recording cache misses? It might be helpful to see that information over time.

Comment (brianosaurus) [reply commons/types/rate_limit.go]: Well, the ratio of hits and misses.

File: commons/types/window.go [hunk 2]
```diff
@@ -125,3 +126,72 @@ func (this Window) ToPrettyJson() string {
...
+	return nil
+}
+
+func (this Window) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		Id        int64				`json:"hash"`
+		Sum       uint64			`json:"hash"`
+		Entries   int64				`json:"hash"`
+		Slope     float64			`json:"hash,omitempty"`
+		TTL       time.Duration		`json:"hash,omitempty"`
```
Comment (brianosaurus) [inline commons/types/window.go]: gofmt

File: commons/utils/linear_regression.go [hunk 3]
```diff
@@ -21,6 +21,9 @@ func LinearRegression(points *[]Point) (a float64, b float64) {
 	}
 
 	base := (n*sumXX - sumX*sumX)
+	if base == 0 {
```
Comment (brianosaurus) [inline commons/utils/linear_regression.go]: Nice catch.

File: dapos/dapos_handshake.go [hunk 4]
```diff
@@ -518,25 +517,33 @@ func executeTransaction(transaction *types.Transaction, receipt *types.Receipt,
...
+
 		err = processDVMResult(transaction, dvmResult, receipt)
 		if err != nil {
 			utils.Error(err)
+
 			receipt.Status = types.StatusInternalError
 			receipt.HumanReadableStatus = err.Error()
 			receipt.Cache(services.GetCache())
+
+			rateLimit, err := types.NewRateLimit(transaction.From, transaction.Hash,  hertz)
```
Comment (brianosaurus) [inline dapos/dapos_handshake.go]: @restevens402 is transaction.From going away? Or do we populate it automatically when we unravel the signature?

File: commons/types/rate_limit.go (hunk 1 above)
Comment (restevens402) [reply commons/types/rate_limit.go]: No, we do not at this time

File: dapos/dapos_handshake.go (hunk 4 above)
Comment (restevens402) [reply dapos/dapos_handshake.go]: right now it is part of the original submitted tx.  We have proposed the plan of eliminating that and using the value derived from the signature.

**Pull request merged**